packages/editor/src/lib/components/Shape.tsx

```diff
@@ -132,23 +143,30 @@ export const Shape = memo(function Shape({
 					className="tl-shape tl-shape-background"
 					data-shape-type={shape.type}
 					draggable={false}
+					style={{ display: isCulled ? 'none' : undefined }}
 				>
-					{isCulled ? null : (
-						<OptionalErrorBoundary fallback={ShapeErrorFallback} onError={annotateError}>
-							<InnerShapeBackground shape={shape} util={util} />
-						</OptionalErrorBoundary>
-					)}
+					<OptionalErrorBoundary fallback={ShapeErrorFallback} onError={annotateError}>
+						<InnerShapeBackground shape={shape} util={util} />
+					</OptionalErrorBoundary>
 				</div>
 			)}
-			<div ref={containerRef} className="tl-shape" data-shape-type={shape.type} draggable={false}>
-				{isCulled ? (
-					<CulledShape shapeId={shape.id} />
-				) : (
-					<OptionalErrorBoundary fallback={ShapeErrorFallback as any} onError={annotateError}>
-						<InnerShape shape={shape} util={util} />
-					</OptionalErrorBoundary>
-				)}
+			<div
+				ref={containerRef}
+				className="tl-shape"
+				data-shape-type={shape.type}
+				draggable={false}
+				style={{ display: isCulled ? 'none' : undefined }}
+			>
+				<OptionalErrorBoundary fallback={ShapeErrorFallback as any} onError={annotateError}>
+					<InnerShape shape={shape} util={util} />
+				</OptionalErrorBoundary>
 			</div>
+			<div
+				ref={culledContainerRef}
+				className="tl-shape__culled"
+				draggable={false}
+				style={{ display: isCulled ? undefined : 'none' }}
```

Comment:
nonblocking: I imagine we could further reduce the culling/unculling overhead by setting the display properties sneakily in our useQuickReactor effect, thus avoiding the react render cycle.

Reply:
couldn't help myself, POC here https://github.com/tldraw/tldraw/pull/3340 seems fast!